awx/main/access.py

```diff
@@ -2407,7 +2407,8 @@ def filtered_queryset(self):
 
     def get_queryset(self):
         return super(UnifiedJobAccess, self).get_queryset().filter(
-            workflowapproval__isnull=True)
+            workflowapproval__isnull=True
+        ).exclude(launch_type='sync')
```

Comment:
I'm not sure why, but doing it this way is a *lot* faster:

![image](https://user-images.githubusercontent.com/214912/64654993-e511a600-d3f8-11e9-92fd-4b5659049462.png)

![image](https://user-images.githubusercontent.com/214912/64655004-e8a52d00-d3f8-11e9-8c8d-4b1ab6a1f8f1.png)

Reply:
If you *do* want to itemize these, you can still do so via:

`/api/v2/project_updates/?launch_type=sync`

Reply:
It's almost certainly slower filtering via the queryset because we add a `.distinct()` in the filters. This was a major topic in the 3.0 RBAC introduction.

https://github.com/ansible/awx/blob/6da445f7c06e1b7babdc20928492f1b0f7c59618/awx/api/filters.py#L335

We still probably have some bugs with duplicated entries in sublists today. They were so hard to get rid of, that basically in filters, we apply a distinct for _any_ sort of filter, and for all filters.

It's probably the case that the filters could be made to selectively apply the distinct as a general solution. The rule is pretty simply that you can filter based on any direct field or forward ForeignKey without risk of incurring duplicates. Reverse one-to-many and many-to-many kind of logically can create duplicates wherever there is more than 1 reference to an object from the related model.

I'm very nervous about filtering them out from the list here.

Reply:
I wonder if there's _anything_ we can do like this.  In my testing on larger tables, it cuts the `/api/v2/unified_jobs`. page load down by 75%.

Reply:
Yes, that's what I'm saying, I really believe we can get the same performance gain by addressing it in the place in filters that's causing a problem and not having entries appear one place and mysteriously disappear in another place.

I mean, `launch_type` is a Char field. It can't possibly be that hard to conditionally not apply that .distinct when you are applying a filter for that type of field. Even if you filter sync jobs like this, it only addresses the slowdown for the dashboard, and query param filters will still be used in other places, adding that nasty distinct on the massive data set you're interested in.

Reply:
Well, filtering it this way addresses it for `/api/v2/unified_jobs`, specifically, which is where the *real* pain point is due to our usage of polymorphism.

Reply:
```diff
diff --git a/awx/api/filters.py b/awx/api/filters.py
index 547d1b3a1a..c538afe92d 100644
--- a/awx/api/filters.py
+++ b/awx/api/filters.py
@@ -9,7 +9,7 @@ from functools import reduce
 # Django
 from django.core.exceptions import FieldError, ValidationError
 from django.db import models
-from django.db.models import Q
+from django.db.models import Q, CharField, IntegerField
 from django.db.models.fields import FieldDoesNotExist
 from django.db.models.fields.related import ForeignObjectRel, ManyToManyField, ForeignKey
 from django.contrib.contenttypes.models import ContentType
@@ -184,6 +184,12 @@ class FieldLookupBackend(BaseFilterBackend):
 
         field, new_lookup = self.get_field_from_lookup(model, lookup)
 
+        print(field)
+        if isinstance(field, (CharField, IntegerField)):
+            needs_distinct = False
+        else:
+            needs_distinct = True
+
         # Type names are stored without underscores internally, but are presented and
         # and serialized over the API containing underscores so we remove `_`
         # for polymorphic_ctype__model lookups.
@@ -214,7 +220,7 @@ class FieldLookupBackend(BaseFilterBackend):
             return value, new_lookups
         else:
             value = self.value_to_python_for_field(field, value)
-        return value, new_lookup
+        return value, new_lookup, needs_distinct
 
     def filter_queryset(self, request, queryset, view):
         try:
@@ -256,7 +262,7 @@ class FieldLookupBackend(BaseFilterBackend):
                         search_filter_relation = 'AND'
                         values = reduce(lambda list1, list2: list1 + list2, [i.split(',') for i in values])
                     for value in values:
-                        search_value, new_keys = self.value_to_python(queryset.model, key, force_text(value))
+                        search_value, new_keys, _ = self.value_to_python(queryset.model, key, force_text(value))
                         assert isinstance(new_keys, list)
                         search_filters[search_value] = new_keys
                     continue
@@ -279,10 +285,13 @@ class FieldLookupBackend(BaseFilterBackend):
                     q_not = True
 
                 # Convert value(s) to python and add to the appropriate list.
+                needs_distinct = False
                 for value in values:
                     if q_int:
                         value = int(value)
-                    value, new_key = self.value_to_python(queryset.model, key, value)
+                    value, new_key, distinct = self.value_to_python(queryset.model, key, value)
+                    if distinct:
+                        needs_distinct = True
                     if q_chain:
                         chain_filters.append((q_not, new_key, value))
                     elif q_or:
@@ -332,7 +341,10 @@ class FieldLookupBackend(BaseFilterBackend):
                     else:
                         q = Q(**{k:v})
                     queryset = queryset.filter(q)
-                queryset = queryset.filter(*args).distinct()
+                print('where distinct would have been applied')
+                if needs_distinct:
+                    print('apply distinct')
+                    queryset = queryset.filter(*args).distinct()
             return queryset
         except (FieldError, FieldDoesNotExist, ValueError, TypeError) as e:
             raise ParseError(e.args[0])
```

If I was reading code that a human wrote, and it had `.exclude(launch_type="sync").distinct()`, then I would request that it be changed (note, you naturally did not write it that way here). This is just obfuscated because it is code written by code. It was a hack that we've been carrying around for a long time. It was fully understood when the hack was applied that it might cause us trouble at some point in the future and then it would have to be revisited.

Reply:
Okay, @AlanCoding, this makes sense.  I've given your change a try _instead of_ my default `.exclude()`:

Before (`devel`):
<img width="592" alt="image" src="https://user-images.githubusercontent.com/214912/64698412-d7980280-d470-11e9-8dd3-7b45d723df44.png">

After (this PR with your suggestion):
<img width="592" alt="image" src="https://user-images.githubusercontent.com/214912/64698462-f0081d00-d470-11e9-9fed-7a0acdd58981.png">

I like this and I'm gonna add it to this PR and see if integration still passes.

Reply:
Actually, I discussed this with @AlanCoding and I'm a little nervous, because I don't think it's going to properly filter out duplicates (I'm seeing evidence that it doesn't).

Reply:
Detecting when `distinct()` is needed is *really* tough. When I went through our endpoints a few years ago trying to figure out if I could remove them.... there are too many places where joins can come in on the queryset... you just can't apply this as part of the filter backend and expect it to work right. You'd need to handle it right before the query is ready to execute.

Reply:
@AlanCoding I want to understand your concern which this change - is it *solely* that removing sync project updates from `/api/v2/unified_jobs/` is a breaking change/not very clear for API users?  That I can understand.

My main thing I'm trying to avoid here is that for users w/ 500k+ jobs in the table the current approach of `/api/v2/unified_jobs/?not__launch_type=sync` is resulting in 5+s queries (which my suggested change here alleviates).

Reply:
> is it solely that removing sync project updates from /api/v2/unified_jobs/ is a breaking change/not very clear for API users? That I can understand.

Yes. Confusing and breaking are fair to summarize.

I totally understand the need for hacks to move on, but I do want to get a reading if people would agree with the general idea of making a later patch to be more selective about using distincts in the filter backend.

Reply:
I don't think anybody is _against_ a more generalized improvement to our usage of `.distinct()`, I've just sensed hesitancy about tackling it _now_, given how complex it might get cc @matburt

Reply:
I'm leery about API changes outside of the change to the deprecated /api/v2/dashboard